src/pypi_attestations/_impl.py

```diff
@@ -331,3 +331,85 @@ def _ultranormalize_dist_filename(dist: str) -> str:
         return f"{name}-{ver}.tar.gz"
     else:
         raise ValueError(f"unknown distribution format: {dist}")
+
+
+class Publisher(BaseModel):
+    """Publisher as defined in PEP 740."""
+
+    kind: str
+    """
+    The kind of Trusted Publisher.
+    """
+
+    claims: Union[object, None] = Field(...)  # noqa: UP007
+    """
+    Claims specified by the publisher.
+    """
+
+    @classmethod
+    def from_kind(cls, kind: str) -> Publisher:
+        """Construct a Publisher from a kind."""
+        return Publisher(kind=kind, claims=None)
+
+
+class AttestationBundle(BaseModel):
+    """AttestationBundle object as defined in PEP 740."""
+
+    publisher: Publisher
+    """
+    The publisher associated with this set of attestations.
+    """
+
+    attestations: list[Attestation]
+    """
+    The list of attestations included in this bundle.
+    """
+
+
+class Provenance(BaseModel):
+    """Provenance object as defined in PEP 740."""
+
+    version: Literal[1]
+    """
+    The provenance object's version, which is always 1.
+    """
+
+    attestation_bundles: list[AttestationBundle]
+    """
+    One or more attestation "bundles".
+    """
+
+
+class ProvenanceError(AttestationError):
+    """The Provenance object was not constructed."""
+
+    def __init__(self: ProvenanceError, msg: str) -> None:
+        """Initialize an `ProvenanceError`."""
+        super().__init__(f"Provenance object not constructed: {msg}")
+
+
+def construct_simple_provenance_object(kind: str, attestations: list[str | bytes]) -> Provenance:
```

Comment:
This could be a `@classmethod` in `Provenance`, similar to `Attestation::from_bundle()`